### ISSUE TITLE
Need to retrieve truncated message

### ISSUE THREAD

**Issue by paulpozicheun** [dsuch/pymqi#148]
Hello,

I would like to retrieve truncated messages from MQ, along with the information of the real size of the message.
The actual get() method from class Queue doesn't fit my need because I can't retrieve the size.
Would it be possible for you to add the following method into file https://github.com/dsuch/pymqi/blob/master/code/pymqi/__init__.py ?

```python
    def get_truncated(self, msgLength, *opts):
        """get_truncated(msgLength [, m_desc, get_opts])

        Return a truncated message from the queue and the real length of the message.
        If the queue is not already open, it is opened now with the option 'MQOO_INPUT_AS_Q_DEF'.

        msgLength specifies the part of the message to be retrieved and is mandatory.

        m_desc is the pymqi.md() MQMD Message Descriptor for receiving
        the message. If it is not passed, or is None, then a default
        md() object is used.

        get_opts is the pymqi.gmo() MQGMO Get Message Options
        structure for the get call. If it is not passed, or is None,
        then a default gmo() object is used.
        To avoid the raise of an exception, option 'MQGMO_ACCEPT_TRUNCATED_MSG'
        will be forced in the get_opts structure.

        If m_desc and/or get_opts arguments were supplied, they may be
        updated by the get operation."""

        m_desc, get_opts = common_q_args(*opts)
        if get_opts is None:
            get_opts = gmo()
        get_opts.Options |= CMQC.MQGMO_ACCEPT_TRUNCATED_MSG
        # If queue open was deferred, open it for put now
        if not self.__qHandle:
            self.__openOpts = CMQC.MQOO_INPUT_AS_Q_DEF
            self.__realOpen()

        # Check message length is not too big
        if msgLength > 4096:
            length = 4096
        else:
            length = msgLength

        rv = pymqe.MQGET(self.__qMgr.getHandle(), self.__qHandle,
                         m_desc.pack(), get_opts.pack(), length)

        # If message was small enough to fit in the buffer => OK
        # If message was too long for the buffer and we get a warning => OK
        # Otherwise => error
        if not rv[-2] or (rv[-2] == CMQC.MQCC_WARNING and rv[-1] == CMQC.MQRC_TRUNCATED_MSG_ACCEPTED):
            m_desc.unpack(rv[1])
            get_opts.unpack(rv[2])
            return rv[0], rv[-3]

        raise MQMIError(rv[-2], rv[-1])
```

Kind regards,

Paul

**Comment by SeyfSV** [comment]
Hi @paulpozicheun,
Yes, you right, there is no way to get the real size of the message. But adding `get_truncated` function is not good way, imho.
What do you think about this way:
```
def get(self, maxLength=None, *opts, force_truncated=True):
        """get([maxLength [, m_desc, get_opts]])

        Return a message from the queue. If the queue is not already
        open, it is opened now with the option 'MQOO_INPUT_AS_Q_DEF'.

        maxLength, if present, specifies the maximum length for the
        message. If the message received exceeds maxLength, then the
        behavior is as defined by MQI and the get_opts argument.

        If maxLength is not specified, or is None, then the entire
        message is returned regardless of its size. This may require
        multiple calls to the underlying MQGET API.

        m_desc is the pymqi.md() MQMD Message Descriptor for receiving
        the message. If it is not passed, or is None, then a default
        md() object is used.

        get_opts is the pymqi.gmo() MQGMO Get Message Options
        structure for the get call. If it is not passed, or is None,
        then a default gmo() object is used.

        If m_desc and/or get_opts arguments were supplied, they may be
        updated by the get operation."""

        m_desc, get_opts = common_q_args(*opts)
        if get_opts is None:
            get_opts = gmo()
        # If queue open was deferred, open it for put now
        if not self.__qHandle:
            self.__openOpts = CMQC.MQOO_INPUT_AS_Q_DEF
            self.__realOpen()

        # Truncated message fix thanks to Maas-Maarten Zeeman
        if maxLength is None:
            length = 4096
        else:
            length = maxLength

        rv = pymqe.MQGET(self.__qMgr.getHandle(), self.__qHandle,
                         m_desc.pack(), get_opts.pack(), length)

        if not rv[-2]:
            # Everything A OK
            m_desc.unpack(rv[1])
            get_opts.unpack(rv[2])
            return rv[0]

        m_desc.unpack(rv[1])  # save the message id
        get_opts.unpack(rv[2])
        length = rv[-3]

        # Process truncated message
        if (rv[-1] == CMQC.MQRC_TRUNCATED_MSG_ACCEPTED or
            # Do not reread message with original length
            (rv[-1] == CMQC.MQRC_TRUNCATED_MSG_FAILED and not force_truncated) or
            # Default buffer used, if not truncated, give up
            (rv[-1] != CMQC.MQRC_TRUNCATED_MSG_FAILED and force_truncated)):
            raise MQMIError(rv[-2], rv[-1],
                            message=rv[0],
                            original_length=rv[-3])

        # Message truncated, but we know its size. Do another MQGET
        # to retrieve it from the queue.
        rv = pymqe.MQGET(self.__qMgr.getHandle(), self.__qHandle,
                         m_desc.pack(), get_opts.pack(), length)
        if rv[-2]:
            raise MQMIError(rv[-2], rv[-1])
        m_desc.unpack(rv[1])
        get_opts.unpack(rv[2])

        return rv[0]
```
and now you can get message size and decide leave the message in the queue or not:
```
import pymqi


broker = '127.0.0.1'
name = 'MQTEST'
channelName = 'CH1'
port = 8887

connection_name = "{0}({1})".format(broker, port)

qm = pymqi.QueueManager(None)
cd = pymqi.CD()
cd.ChannelName = channelName.encode()
cd.ConnectionName = connection_name.encode()
cd.ChannelType = pymqi.CMQC.MQCHT_CLNTCONN
cd.TransportType = pymqi.CMQC.MQXPT_TCP
connectOptions = pymqi.CMQC.MQCNO_HANDLE_SHARE_BLOCK
qm.connectWithOptions(name, cd=cd, opts=connectOptions)


if __name__ == '__main__':
    try:
        queue = pymqi.Queue(qm, 'TEST.QUEUE', pymqi.CMQC.MQOO_INPUT_AS_Q_DEF | pymqi.CMQC.MQOO_OUTPUT)
        md_put = pymqi.MD()
        queue.put(b'12345', md_put)
        gmo = pymqi.GMO()
        gmo.Options = pymqi.CMQC.MQGMO_ACCEPT_TRUNCATED_MSG # Comment to leave message in queue
        gmo.MatchOptions = pymqi.CMQC.MQMO_MATCH_MSG_ID

        md_get = pymqi.MD()
        md_get.MsgId = md_put.MsgId

        message = queue.get(1, md_get, gmo, force_truncated=False) # force_truncated=True to read complete message, it is default behaviour
        print('Complete message read from queue')
        print(message)


    except pymqi.MQMIError as ex:
        if ex.comp == pymqi.CMQC.MQCC_WARNING:
            if ex.reason == pymqi.CMQC.MQRC_TRUNCATED_MSG_ACCEPTED:
                print('Truncated message read from queue')
                print(ex.message)
                print(ex.original_length)
            elif ex.reason == pymqi.CMQC.MQRC_TRUNCATED_MSG_FAILED:
                print('Truncated messge did not read from queue')
                print(ex.message)
                print(ex.original_length)
    except Exception as ex:
        print(ex)
    finally:
        qm.disconnect()
```

**Comment by paulpozicheun** [comment]
Hello @SeyfSV ,

I've tested your new get() method and the way to handle a truncated message,
and it worked perfectly for me.

I do agree that there is no need to create a new function.

Can you integrate your changes ?
Can we wait for the new commit to close this issue ?

Thanks a lot for the job.

Kind regards,
Paul

**Comment by SeyfSV** [comment]
Hello @paulpozicheun, I committed changes in separate branch (b89ba91e8c1249845b2c519ba23e86b18069bfb1) for @dsuch review.
I modified get method and removed `force_truncated` attribute. Now it looks as before and work exactly as written in method's docstring.
Also I added tests by which you can check behaviour.

I or @dsuch will close issue after merging branch.

**Comment by SeyfSV** [comment]
Fixed in 2e3d191aed95ca928e19d756b9a8ce876e3b35c1
